core/query/evalMixingQuery.ml

```diff
@@ -0,0 +1,121 @@
+(*****************************************************************************
+ ** simpleSqlGen.ml - Simple Sql syntax generator                           **
+ **                                                                         **
+ ** author: Wilmer Ricciotti                                                **
+ *****************************************************************************)
+
+open Utility
+open CommonTypes
+
+module Q = MixingQuery.Lang
+module E = MixingQuery.Eval
+module C = Constant
+module S = Sql
+
+
+exception EvalMixingUnimplemented of string
+
+let error msg : 'a = raise (EvalMixingUnimplemented msg)
+let eval_error fmt : 'r =
+    Printf.ksprintf error fmt
+
+let mapstrcat sep f l = l |> List.map f |> String.concat sep
+
+let dummy_sql_empty_query =
+    (false,S.Fields [(S.Constant (Constant.Int 42), "@unit@")], [], S.Constant (Constant.Bool false), [])
+
+(* convert an NRC-style query into an SQL-style query *)
+let rec sql_of_query is_set = function
+| Q.Concat ds -> S.Union (is_set, List.map (disjunct is_set) ds, 0)
+| q -> disjunct is_set q
+
+and disjunct is_set = function
+| Q.Prom p -> sql_of_query true p
+| Q.Singleton _ as j -> S.Select (body is_set [] [] j)
+| Q.For (_, gs, os, j) -> S.Select (body is_set gs os j)
+| _arg -> Debug.print ("error in SimpleSqlGen.disjunct: unexpected arg = " ^ Q.show _arg); failwith "disjunct"
+
+and generator locvars = function
+| (v, Q.Prom p) -> (S.Subquery (E.contains_free locvars p, sql_of_query true p, v))
+| (v, Q.Table (_, tname, _, _)) -> (S.TableRef (tname, v))
+| (v, Q.Dedup (Q.Table (_, tname, _, _))) ->
+    S.Subquery (false, S.Select (true, S.Star, [S.TableRef (tname, v)], S.Constant (Constant.Bool true), []), v)
+| (_, _arg) -> Debug.print ("error in SimpleSqlGen.disjunct: unexpected arg = " ^ Q.show _arg); failwith "generator"
+
+and body is_set gs os j =
+    let selquery body where =
+        let froms =
+            gs
+            |> List.fold_left (fun (locvars,acc) (v,_q as g) -> (v::locvars, generator locvars g::acc)) ([],[])
+            |> snd
+            |> List.rev
+        in
+        let os = List.map base_exp os in
+        (is_set, S.Fields body, froms, where, os)
+    in
+    match j with
+    | Q.Concat [] -> dummy_sql_empty_query
+    | Q.Singleton (Q.Record fields) ->
+        selquery
+        <| List.map (fun (f,x) -> (base_exp x, f)) (StringMap.to_alist fields)
+        <| Sql.Constant (Constant.Bool true)
+    | Q.If (c, Q.Singleton (Q.Record fields), Q.Concat []) ->
+        selquery
+        <| List.map (fun (f,x) -> (base_exp x, f)) (StringMap.to_alist fields)
+        <| base_exp c
+    | _ -> Debug.print ("error in SimpleSqlGen.body: unexpected j = " ^ Q.show j); failwith "body"
+
+and base_exp = function
+(* XXX: Project expects a (numbered) var, but we have a table name
```

Comment:
Perhaps document this expectation executably with an assertion or error message

Reply:
Actually I don't think a projection from a table would be well-typed - projection in NRC queries takes a record, not a list of records (or a table, which is a different type from a list of records).  So probably safe to leave out this specific error but OK to leave it too.

Reply:
Yeah, I guess there was no need to be worried about it in the first place. I probably wasn't thinking straight when I wrote that comment. I'm just a bit hesitant to remove it because of the remote possibility that last year I saw something that I'm not seeing now.